index.js

```diff
@@ -1,33 +1,81 @@
+};
+
+const defaultFooterResolver = (currentPageIndex, pagesLength) =>`Page ${currentPageIndex + 1} / ${pagesLength}`;
+
+/**
+ *
+ * @param {Message} msg - the message
+ * @param {MessageEmbed[]} pages - array of message embeds to use as each page.
+ * @param {PaginationOptions} paginationOptions - exposes collector options, provides customization.
+ */
+const paginationEmbed = async (
+										msg, pages,
+										{
+											deleteOnEnd = false, useUtil = false,
+											emojiList = ['⏪', '⏩'],
+											collectorFilter = null,
+											footerResolver = defaultFooterResolver,
+											pageResolver = defaultPageResolver,
+											timeout = 120000,
+											...rest
+										} = {}) => {
 	if (!msg && !msg.channel) throw new Error('Channel is inaccessible.');
 	if (!pages) throw new Error('Pages are not given.');
-	if (emojiList.length !== 2) throw new Error('Need two emojis.');
 	let page = 0;
-	const curPage = await msg.channel.send(pages[page].setFooter(`Page ${page + 1} / ${pages.length}`));
-	for (const emoji of emojiList) await curPage.react(emoji);
+	pages[page].setFooter(footerResolver(page, pages.length));
+	const curPage = useUtil ? await msg.util.reply(pages[page]) : await msg.channel.send(pages[page]);
 	const reactionCollector = curPage.createReactionCollector(
-		(reaction, user) => emojiList.includes(reaction.emoji.name) && !user.bot,
-		{ time: timeout }
+		(reaction, user) => emojiList.includes(reaction.emoji.name) && !user.bot
+							&& (collectorFilter ? collectorFilter(reaction, user, emojiList) : true),
+		{ time: timeout, ...rest }
 	);
-	reactionCollector.on('collect', reaction => {
-		reaction.users.remove(msg.author);
-		switch (reaction.emoji.name) {
-			case emojiList[0]:
-				page = page > 0 ? --page : pages.length - 1;
-				break;
-			case emojiList[1]:
-				page = page + 1 < pages.length ? ++page : 0;
-				break;
-			default:
-				break;
-		}
-		curPage.edit(pages[page].setFooter(`Page ${page + 1} / ${pages.length}`));
+	reactionCollector.on('collect', async (reaction, user) => {
+		// In case a react deletes the curPage msg in the pageResolver and another react is quickly made.
+		if (curPage.deleted) return;
+		await reaction.users.remove(user.id);
+		const currentPage = page;
+		page = await pageResolver(curPage, pages, emojiList, page, reaction);
+
+		if ( !curPage.deleted && currentPage != page && page >= 0 && page < pages.length)
+			curPage.edit(pages[page].setFooter(footerResolver(page, pages.length)));
 	});
-	reactionCollector.on('end', () => {
+	reactionCollector.on('end', async () => {
 		if (!curPage.deleted) {
-			curPage.reactions.removeAll()
+			if (deleteOnEnd)
+				await curPage.delete();
+			else
+				curPage.reactions.removeAll();
 		}
 	});
+	// This will fail if the message is deleted before the reactions are all added, bail before reactor creation
+	try {
+		for (const emoji of emojiList) await curPage.react(emoji);
+	} catch (error) {
+			reactionCollector.stop();
```

Comment:
The reaction collector will automatically stop if the message it references was deleted, so there's no need for an explicit `try`/`catch` here; just attempt to add the reactions and let any `10008` unknown message errors propagate as usual.